### ISSUE TITLE
mergeGraph in the graph DSL does not incorporate edges from g1

### ISSUE THREAD

**Issue by yihozhang** [MrVPlusOne/Singularity#2]
Hi,

In [StandardSystem.scala](https://github.com/MrVPlusOne/Singularity/blob/develop/src/main/scala/singularity/StandardSystem.scala#L452) line 452, the mergeGraph operation seems only to include edges from g2, but not g1, which is confusing and may probably be wrong. I haven't done any tests on performance yet so feel free to close this issue if this is on purpose.

```scala
    val mergeGraph = mkAbstract("mergeGraph", tyVarNum = 1,
      typeInstantiation = {
        case IS(eT) => IS(EGraph(eT), EGraph(eT)) -> EGraph(eT)
      }, eval = {
        case IS(g1: GraphValue, g2: GraphValue) =>
          GraphValue(g1.nodeNum + g2.nodeNum, g2.shiftIndex(g1.nodeNum).edges)
      })
```

Although I believe (from the commit history) this repo is no longer maintained, I found such typo in the codebase could potentially influence the performance of the genetic programming algorithm, and fixing it may help future users better utilize the power of Singularity.

**Comment by MrVPlusOne** [comment]
Hi Yihong,

You are absolutely right! I think the right expression there should be `g1.edges + g2.shiftIndex(g1.nodeNum).edges`. Feel free to send a pull request, or I will fix it later.

Thanks,
Jiayi

**Comment by yihozhang** [comment]
Hi Jiayi,

Thanks for the quick reply! The proposed expression looks good to me. It'd be great if you could fix it as my local clone of the repo is being messed up :),

As a side note, I found evolutionary training and extrapolation will report OutOfMemory error on some large valueOfInterest (say, fuzzing a graph with 100,000 nodes), which makes the tool currently hard to scale to large input generations. Though haven't had my hypothesis fully tested yet, I suspect this may be caused by the memoization machinery of Stream when creating consecutive patterns induced by RCGs. Here, memoization may cause memory leak as it holds the whole sequence of the output patterns while only the last few are needed. In the example of a graph with 100,000 nodes, the stream may try to store asymptotically 100,000 such graphs, which is very costly. A potential fix may be to use Iterator instead of Stream. Have you encountered similar situations before? A fix will be very appreciated but I also respect your developer time so I'm totally ok with it if you choose not to address it.

Best,
Yihong

**Comment by MrVPlusOne** [comment]
Hi Yihong,

Yes, we did encounter similar issues when trying to generate very large inputs from the found pattern. This is an issue that's somewhat inherent to Singularity's immutable DSL, which contains operations that can be quite expensive to execute for large inputs. 

We were using `Stream` because we wanted to keep not just the last input, but also some previous inputs for other purposes (such as pattern inspection/cost fitting, etc), and the code was not heavily optimized for generating very large inputs. For some patterns that are reusing most of the immutable data structures between iterations, using either `Stream` or `Iterator` shouldn't matter; for others, using `Iterator` might save you an extra `O(n)` factor of memory usage. I currently don't have much time to work on this as I'm occupied by other projects, but I would encourage you to experiment with it and make any necessary changes if generating large inputs is your priority :) In the past, when we needed to generate really large examples, we simply run Singularity at a much smaller `sizeOfInterest` and manually rewrote the pattern into more efficient code using mutable data structures.

Best,
Jiayi
